sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala

```diff
@@ -169,7 +182,13 @@ class DynamicPartitionDataWriter(
 
   /** Extracts the partition values out of an input row. */
   private lazy val getPartitionValues: InternalRow => UnsafeRow = {
-    val proj = UnsafeProjection.create(description.partitionColumns, description.allColumns)
+    val partitionExpression =
+      toBoundExprs(description.partitionColumns, description.allColumns).map {
+        case e: Expression if e.dataType == StringType =>
+          Empty2Null(e)
```

Comment:
We need to do it earlier. In `FileFormatWriter.write`, we sort the input RDD by partition columns, we need to normalize partition values before sorting.

Reply:
@cloud-fan Thanks for review, I have moved it before sort, `PartitionColumns` isretained because it is used to calculate `getPartitionPath`